mono/mini/method-to-ir.c

```diff
@@ -2246,41 +2246,39 @@ test_tailcall (MonoCompile *cfg, MonoBoolean tailcall)
 
 static MonoCallInst *
 mono_emit_call_args (MonoCompile *cfg, MonoMethodSignature *sig, 
-		     MonoInst **args, gboolean calli, gboolean virtual_, gboolean *ptailcall,
```

Comment:
Wouldn't be better to do the cfg->check_pinvoke_callconv check in CEE_CALLI  so tailcall doesn't have to be an out argument anymore ? These checks should be done on an upper level and not emit_calli which is a low level function.

Reply:
I was thinking the opposite -- put them lower -- but put them in one place.
Maybe. You are right if they are earlier/higher than it doesn't change as much and doesn't have to be passed back out. Blocking? I'm debugging other related problems.. but I'll look..

Reply:
Maybe do it in is_supported_tail_call? Still a bit disjointed, because that means is_supported_tail_call understands the implication that check_sp is needed.

Reply:
Not so simple, because the decision of call vs. calli is spread around.

Reply:
One way, maybe gross, to bottleneck it, is is_supported_tail_call return bitfields -- is call supported, is calli supported..

Reply:
mono_emit_call_args was already making a decision, based on llvm_only, and moving this up..does it apply across every single call?

Reply:
The out parameter as I did it, keeps various decision making where it was, and avoids spreading out decisions, except for tailcall decisions. Kinda tough to have it all ways -- can't have everything in one place.

Reply:
is_supported_tail_call () seems like a good place to do it. just add a g_assert (!tailcall) to the 
check_sp code.

Reply:
For check_sp I have that coded..it isn't too too terrible, but slightly bad, put comments in both, and the assert as you say...but what about the llvm_only check?

And is the check_sp logic specific to calli? I think not so ok.

Reply:
To redundantly put it -- mini_emit_calli_full seems ok to fix, maybe, the logic can go in is_tail_call_supported I think. But that still leaves mono_emit_call_args.
It is worse there, because we don't know it is calli currently at that point.

Reply:
emit_call_args () used to disable tail calls because the caller didn't need the information, if it does now, the check should be done in the caller, or the caller's caller. There are only two places which can pass tail==TRUE to that function.

Reply:
I understand what you mean, but it is messy. I believe it is more complicated than just the two places.

check_sp involves duplicating code but I understand what to do.
Because this is about the function being called.

But something like:
mono_emit_call_args:
	if (calli && !cfg->backend->have_op_tail_call_reg)
		no_tailcall = "architecture lacking tail_call_reg";

is messier because higher up we aren't even doing a calli.
But we need/want to omit the return code after the call, that got turned into calli.
When I left in the return code, I hit assert(reg > -1).

I think what we'd have to do, to kinda satisfy you is right before mini_emit_calli_full, we'd call something like is_tail_call_supported, but another variation of it. is_tailcall_supported_calli.

Or in fact, the same function, but add a calli parameter to it, false on the first call, and then later try again with true. Or pass in CEE_CALLI as a close enough approximation.

I can try this but I don't think the result will be particularly cleaner than what I have.